### PR TITLE
feat: add get_keys_by_owner read method

### PR DESCRIPTION
**Pull request by ChaituVR**
## Summary

Adds `get_keys_by_owner`: a public JSON-RPC method that lets users fetch their own API keys, authenticated with their Snapshot alias — the same signing flow the sequencer accepts for follow and vote, so the UI needs no wallet popup once an alias exists.

- Request `{ from, alias, timestamp, sig }` where `sig` is an EIP-712 `GetKeys` signature made by the alias (domain `snapshot 0.1.4`)
- The server recovers the signer, checks it matches `alias`, then confirms on the hub that the alias is registered for `from` and less than 90 days old (same window the sequencer applies)
- Returns the owner's active keys **including the key value** — the read is now owner-only, so the earlier reason to strip it is gone
- Errors: 400 invalid address or signature, 401 alias not registered, 500 hub unreachable
- e2e suite runs against a stub hub started inside the test, so CI has no network dependency
- `generate_key` no longer logs the signature (the key is derived from it)

## How to test

Easiest through the UI: snapshot-labs/sx-monorepo#2247 has step-by-step instructions — run this branch locally and drive it from the settings page.

Or directly: see `get_keys_by_owner` in the README for the message shape and a curl example; sign with any registered alias.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/methods.ts [hunk 1]
```diff
@@ -122,6 +124,62 @@ export const getKeys = async (app: string) => {
...
+    let owner: string;
+    try {
+      owner = getAddress(from);
+    } catch {
+      return { error: 'Invalid address', code: 400 };
+    }
+
+    let signer: string;
+    try {
+      signer = recoverGetKeysSigner({ from, alias, timestamp }, sig);
```
Comment (tony8713) [inline src/methods.ts:155]: `timestamp` is signed but never read. Verified on this branch: `timestamp: 0`, `1`, `now - 3 years` and year 3000 all return 200 with the key value, and re-sending one captured body three times in a row succeeds every time.

So the request blob is a bearer credential for the key value until the alias ages past 90 days, not a per-request proof. A `Math.abs(now - timestamp) > 5 * 60` reject would close it — otherwise the field is decorative and a client author will reasonably assume freshness is enforced because the README documents it.

File: src/methods.ts [hunk 2]
```diff
@@ -122,6 +124,62 @@ export const getKeys = async (app: string) => {
...
+      usageRows.reduce((acc, { key, app, total }) => {
+        (acc[key] ??= []).push({ app, total });
+        return acc;
+      }, {});
+
+    return {
+      keys: rows.map(row => ({ ...row, usage: usageByKey[row.key] ?? [] }))
+    };
+  } catch (err) {
+    capture(err, { context: { params } });
```
Comment (tony8713) [inline src/methods.ts]: `params` still contains `sig` here. Confirmed by pointing `SENTRY_DSN` at a local sink and failing the hub — the envelope carries the full signature at `contexts.context.params.sig`. `snapshot-sentry` only scrubs env values and exception frames, not contexts.

Combined with the missing timestamp check above, that puts a replayable read credential into Sentry on every hub blip. The same applies to `generateKey`'s capture on L75 — that is the sig this PR just removed from the `console.log` seven lines above it, so it is half-fixed. `{ context: { params: { from, alias, timestamp } } }` keeps everything useful for debugging.

File: src/methods.ts [hunk 3]
```diff
@@ -122,6 +124,62 @@ export const getKeys = async (app: string) => {
...
+
+    if (signer !== getAddress(alias))
+      return { error: 'Invalid signature', code: 400 };
+    if (!(await isAliasOf(owner, signer)))
+      return { error: 'Alias not authorized', code: 401 };
+
+    const rows = await db.queryAsync(
+      `
+        SELECT \`key\`, owner, name, tier, created, updated, active
+        FROM \`keys\` WHERE owner = ? AND active = 1`,
```
Comment (tony8713) [inline src/methods.ts]: Missing ``AND `key` IS NOT NULL``. `getActiveKeys` has that guard (L45), and `metrics.ts` L20-22 counts null-key rows as a distinct "pending" status — whitelisted but never generated.

Repro: whitelist an owner without generating, and this returns `{"key": null, "name": "pending", "active": 1, ...}`. Either filter those out or the UI has to special-case a key whose value is `null`.

File: src/methods.ts [hunk 4]
```diff
@@ -122,6 +124,62 @@ export const getKeys = async (app: string) => {
...
+      [owner]
+    );
+    if (!rows.length) return { keys: [] };
+
+    const keyHashes = rows.map(row => row.key).filter(Boolean);
+    const usageRows = keyHashes.length
+      ? await db.queryAsync(
+          `
+        SELECT \`key\`, app, total FROM reqs_monthly
```
Comment (tony8713) [inline src/methods.ts]: Usage does not survive a key rotation. `writer.updateKey` rewrites `keys.key` in place while `reqs_monthly` stays keyed on the old hash, so the counters orphan.

Repro: seed 42 + 7 reqs for the current month, call `generate_key`, and this returns `usage: []`. Anyone who has ever regenerated their key will be shown 0 requests, which reads worse than showing nothing. Not introduced here, but this endpoint is the first thing to surface it to users.

File: src/helpers/eip712.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { verifyTypedData } from '@ethersproject/wallet';
+
+const DOMAIN = { name: 'snapshot', version: '0.1.4' };
+
+const GET_KEYS_TYPES = {
+  GetKeys: [
+    { name: 'from', type: 'string' },
```
Comment (tony8713) [inline src/helpers/eip712.ts]: `from` as `string` diverges from snapshot.js. `aliasTypes` is these exact three fields and uses `from: address`, as do follow/unfollow/subscribe/unsubscribe/profile/statement. The `string` variant is only used by the proposal and vote types, where `from` can be a 32-byte Starknet address.

It costs something here: `string` hashes the literal, so casing has to round-trip exactly. Signing a checksummed `from` and sending it lowercased returns 400, while `alias` in the same struct is `address` and is case-agnostic — two address fields, two different rules. And `getAddress(from)` on L133 rejects Starknet addresses anyway, so the looser type buys nothing.

**Review by tony8713 — COMMENTED** [review]
Read the diff and stood the branch up locally (MySQL 10.11) — `yarn lint`, `yarn typecheck` and `yarn build` are clean, `yarn test:e2e` is 27/27 green. Drove the endpoint against a stub hub to check the edge cases below rather than reading for them; repros are in each thread.

Nothing here is blocking, but (1) and (2) compound: the signed payload is a permanent read credential for the key value, and it is also being written to Sentry.

Three smaller things not worth their own thread:

- `isAliasOf` maps a hub 200 with an unexpected body (`{"data":null}`, `{}`) to `401 Alias not authorized` rather than a 5xx — verified. During a hub degradation, users get told to re-register an alias that is fine.
- This is the first public method that makes an outbound call (`generate_key` only touches the local DB). No rate limit, and CORS is `*`, so any caller with a throwaway wallet can fan out hub requests. Probably fine behind the 3s timeout, just flagging it.
- Conflicts with #95, which rewrites this same region of `methods.ts` and deletes `helpers/mysql.ts`. More subtly: #95 turns `created`/`updated` into `timestamptz` and `key` into `NOT NULL`. Because the response spreads `...row` straight out, the public shape of this new method flips from unix ints to ISO strings depending on merge order. Worth pinning the response shape explicitly here rather than inheriting it from the column types.

Checked and fine, for the record: the hub's `where: { address, alias }` is case-insensitive on both (tested against hub.snapshot.org with all four casing combinations), so no checksum regression; `created_gt` is a real `Int` filter on `AliasWhere`; 90 days does match the live sequencer (`sx-monorepo/apps/sequencer`, #2115 — the 30-day `snapshot-sequencer` repo is archived); cheap checks are correctly ordered before the hub call; and there is no cross-owner leak — an alias registered for A asking for B gets a 401.

File: src/methods.ts [hunk 6]
```diff
@@ -122,6 +125,52 @@ export const getKeys = async (app: string) => {
...
+    try {
+      signer = recoverGetKeysSigner({ from, alias, timestamp }, sig);
+    } catch {
+      return { error: 'Invalid signature', code: 400 };
+    }
+
+    if (signer !== getAddress(alias))
+      return { error: 'Invalid signature', code: 400 };
+    if (!(await isAliasOf(owner, signer)))
+      return { error: 'Alias not authorized', code: 401 };
```
Comment (wa0x6e) [inline src/methods.ts:163]: The hub round-trip can't change the outcome when `from` has no row in `keys` — the query below returns `{keys: []}` either way. Gating on the existing `isWhitelist` (a PK point lookup) skips it, and `401 Not whitelisted` is truer than `200 {keys: []}`, matching what `generate_key` already returns for that state.

```suggestion
    if (signer !== getAddress(alias))
      return { error: 'Invalid signature', code: 400 };
    if (!(await isWhitelist(owner)))
      return { error: 'Not whitelisted', code: 401 };
    if (!(await isAliasOf(owner, signer)))
      return { error: 'Alias not authorized', code: 401 };
```

Two notes:

- This runs before auth, so the two distinct 401s let anyone probe whether an address is a keycard owner. One generic 401 for both closes that.
- Four e2e tests leave `OWNER` un-whitelisted and would hit the new 401 first (empty-list, alias-not-registered, both hub-failure). Whitelisting `OWNER` in `beforeEach` covers three; the empty-list test should deactivate the key instead, so it exercises "whitelisted, no active key".

File: src/helpers/aliases.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      variables: { address, alias, created_gt: createdGt }
+    }),
+    signal: AbortSignal.timeout(HUB_TIMEOUT_MS)
+  });
+  if (!res.ok) throw new Error(`Hub responded with ${res.status}`);
+
+  const body = await res.json();
+  if (!body.data) throw new Error('Hub returned no data');
+  if (body.errors)
+    throw new Error(body.errors[0]?.message || 'Hub returned GraphQL errors');
```
Comment (wa0x6e) [inline src/helpers/aliases.ts]: Wrong order: a GraphQL error-only response is `{"errors": [...]}` with `data` omitted, so `!body.data` fires first and the hub's actual message on the next line is unreachable. Sentry records `Hub returned no data` for every hub-side failure.

```suggestion
  if (body.errors)
    throw new Error(body.errors[0]?.message || 'Hub returned GraphQL errors');
  if (!body.data) throw new Error('Hub returned no data');
```

Diagnostics only — the client gets the same 500 either way, so no test needed (the message never crosses the HTTP boundary).

File: src/methods.ts [hunk 8]
```diff
@@ -122,6 +125,52 @@ export const getKeys = async (app: string) => {
...
+    let owner: string;
+    try {
+      owner = getAddress(from);
+    } catch {
+      return { error: 'Invalid address', code: 400 };
+    }
+
+    const ts = Date.now() / 1e3;
+    if (timestamp > ts + SIGNATURE_WINDOW || timestamp < ts - SIGNATURE_WINDOW)
+      return { error: 'Signature expired', code: 401 };
```
Comment (wa0x6e) [inline src/methods.ts]: `params.timestamp` is used raw, so `undefined`, a non-numeric string or an object make both comparisons `NaN` and the freshness gate silently passes. Not exploitable — ethers then rejects those same values at `uint64` encoding — but the caller gets `400 Invalid signature` when the real problem is the timestamp.

```suggestion
    if (typeof timestamp !== 'number' || !Number.isFinite(timestamp))
      return { error: 'Invalid timestamp', code: 400 };

    const ts = Date.now() / 1e3;
    if (timestamp > ts + SIGNATURE_WINDOW || timestamp < ts - SIGNATURE_WINDOW)
      return { error: 'Signature expired', code: 401 };
```

File: test/e2e/get_keys_by_owner.test.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+          if (hubMode === 'graphql-error') {
+            return res.end(
+              JSON.stringify({ errors: [{ message: 'rate limited' }] })
+            );
+          }
+
+          const { address, alias } = JSON.parse(body).variables;
+          const aliases = registered.has(`${address}:${alias}`.toLowerCase())
+            ? [{ address, alias }]
+            : [];
```
Comment (wa0x6e) [inline test/e2e/get_keys_by_owner.test.ts]: The stub never reads `created_gt`, so the 90-day `ALIAS_EXPIRY_DAYS` cutoff has no coverage. Drop or rename that variable in `aliases.ts` and every test here still passes, while the hub — where `created_gt` is optional — silently starts accepting aliases of any age.

Suggest having the stub filter on it the way the hub does. The default third arg keeps existing `registerAlias` calls unchanged:

```diff
-const registered = new Set<string>();
+const registered = new Map<string, number>();

-function registerAlias(address: string, alias: string) {
-  registered.add(`${address}:${alias}`.toLowerCase());
+function registerAlias(
+  address: string,
+  alias: string,
+  createdAt = Math.floor(Date.now() / 1e3)
+) {
+  registered.set(`${address}:${alias}`.toLowerCase(), createdAt);
 }

-          const { address, alias } = JSON.parse(body).variables;
-          const aliases = registered.has(`${address}:${alias}`.toLowerCase())
-            ? [{ address, alias }]
-            : [];
+          const { address, alias, created_gt } = JSON.parse(body).variables;
+          const createdAt = registered.get(`${address}:${alias}`.toLowerCase());
+          const aliases =
+            createdAt !== undefined && createdAt > created_gt
+              ? [{ address, alias }]
+              : [];
```

plus one test registering at `now - 91 days` and expecting `401 Alias not authorized`.

File: src/methods.ts [hunk 10]
```diff
@@ -122,6 +125,52 @@ export const getKeys = async (app: string) => {
   }
 };
 
+export const getKeysByOwner = async (params: any) => {
+  try {
+    const { from, alias, timestamp, sig } = params;
```
Comment (wa0x6e) [inline src/methods.ts]: `rpc.ts`'s `params = {}` default only substitutes for `undefined`, never `null`. So `{"method":"get_keys_by_owner","params":null}` throws on this line, lands in the outer catch, fires `capture()` and returns 500 — unauthenticated, free and repeatable.

```suggestion
    const { from, alias, timestamp, sig } = params ?? {};
```

Then `getAddress(undefined)` throws into the existing inner catch and returns a clean `400 Invalid address`, matching `generateKey`.

File: src/methods.ts [hunk 11]
```diff
@@ -122,6 +125,52 @@ export const getKeys = async (app: string) => {
...
+
+    let owner: string;
+    try {
+      owner = getAddress(from);
+    } catch {
+      return { error: 'Invalid address', code: 400 };
+    }
+
+    const ts = Date.now() / 1e3;
```
Comment (wa0x6e) [inline src/methods.ts]: Nit: `Date.now()` is integer milliseconds, so this keeps a fractional part — the window ends up ~300.999s on one side and ~299.001s on the other. `aliases.ts:22` floors the same expression.

```suggestion
    const ts = Math.floor(Date.now() / 1e3);
```

Overlaps the timestamp-guard suggestion above, which reproduces this line — take one, not both.

Also: `401 Signature expired` isn't in the PR description's documented error list.

File: src/methods.ts (hunk 2 above)
Comment (wa0x6e) [reply src/methods.ts]: Not a real concern

File: src/methods.ts (hunk 6 above)
Comment (ChaituVR) [reply src/methods.ts:163]: Skipping — `isWhitelist` is just "has a row in `keys`", which stops being a gate once self-serve `create_key` lands, and as you spotted it would sit before `isAliasOf`.

File: src/methods.ts [hunk 12]
```diff
@@ -150,7 +150,7 @@ export const getKeysByOwner = async (params: any) => {
     const rows = await db.queryAsync(
       `
         SELECT \`key\`, owner, name, tier, created, updated, active
-        FROM \`keys\` WHERE owner = ? AND active = 1`,
+        FROM \`keys\` WHERE owner = ? AND active = 1 AND \`key\` IS NOT NULL`,
```
Comment (wa0x6e) [inline src/methods.ts:168]: I think that from a commit around February 2026, `key` can not be null anymore on creation. Any remaining data with null keys are legacy, and could be deleted (never activated for months).

We could delete all rows with null key, update the schema to not allow null, and remove all the null handling from the code

**Review by wa0x6e — APPROVED** [review]
tAck

File: src/methods.ts [hunk 13]
```diff
@@ -122,6 +125,55 @@ export const getKeys = async (app: string) => {
   }
 };
 
+export const getKeysByOwner = async (params: any) => {
+  try {
+    const { from, alias, timestamp, sig } = params ?? {};
```
Comment (Sekhmet) [inline src/methods.ts]: Maybe we could type params there? On call-site params will be any typed so it shouldn't change anything else, but now all those params in this function will have known types.

**Review by Sekhmet — COMMENTED** [review]
One suggestion for improving typing in getKeysByOwner, otherwise it looks good to me.
